testcases/kernel/syscalls/madvise/madvise08.c

```diff
@@ -0,0 +1,297 @@
+/*
+ * Copyright (c) 2016-2017 Richard Palethorpe <richiejp@f-m.fm>
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+/*
+ * Check that memory marked with MADV_DONTDUMP is not included in a core dump
+ * and check that the same memory then marked with MADV_DODUMP is included in
+ * a core dump.
+ *
+ * In order to reliably find the core dump this test temporarily changes the
+ * system wide core_pattern setting. Meaning all core dumps will be sent to the
+ * test's temporary dir untill the setting is restored during cleanup.
+ *
+ * Test flow: map memory,
+ *	      write generated character sequence to memory,
+ *	      start child process,
+ *	      mark memory with MADV_DONTDUMP in child,
+ *	      abort child,
+ *	      scan child's core dump for character sequence,
+ *	      if the sequence is not found it is a pass otherwise a fail,
+ *	      do the opposite for MADV_DODUMP,
+ */
+
+#include "tst_test.h"
+#include <sys/types.h>
+#include <sys/wait.h>
+#include <fcntl.h>
+#include <unistd.h>
+#include <signal.h>
+#include <stdlib.h>
+#include <stdio.h>
+
+#if defined(MADV_DONTDUMP) && defined(MADV_DODUMP)
+#define CORE_PATTERN "/proc/sys/kernel/core_pattern"
+
+struct test_case {
+	int fd;
+	char *fname;
+	char *fcontent;
+	void *fmem;
+	size_t fmemsize;
+	int ycount;
+	char cpattern[1048];
+	ssize_t cpsize;
+	int cpfd;
+	struct stat stat;
+	int dfd;
+} tc;
+
+/*
+ *TODO: Check/set /proc/PID/coredump_filter,
+ *		RLIMIT_CORE, RLIMIT_FSIZE,
+ *		PR_SET_DUMPABLE,
+ *		CONFIG_COREDUMP (kernel)
+ * Currently it is assumed that if the core dump can not be found it is because
+ * they are disabled rather than a bug. Also tests may pass because of the core
+ * and file size limits.
+ */
+
+static void setup(void)
+{
+	const char *dumpf = "dump-%p";
+	char cwd[1024];
+	char tmpcpattern[1048];
+
+	tc.fd = 0;
+	tc.fname = "mmf0";
+	tc.fmem = NULL;
+	tc.cpsize = -1;
+	tc.cpfd = 0;
+	tc.dfd = 0;
+	tc.ycount = 1280;
+	tc.fmemsize = tc.ycount + 2;
+
+	SAFE_GETCWD(cwd, 1024);
+	snprintf(tmpcpattern, 1048, "%s/%s", cwd, dumpf);
+	tst_res(TINFO, "Temporary core pattern is '%s'", tmpcpattern);
+
```

Comment:
We have ```SAFE_FILE_PRINTF()``` which is designed specially for writing a proc files in one function call.